rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp

```diff
@@ -0,0 +1,207 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <chrono>
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rosbag2_compression/sequential_compression_reader.hpp"
+#include "rosbag2_compression/sequential_compression_writer.hpp"
+#include "rosbag2_cpp/converter_options.hpp"
+#include "rosbag2_cpp/readers/sequential_reader.hpp"
+#include "rosbag2_cpp/reader.hpp"
+#include "rosbag2_cpp/storage_options.hpp"
+#include "rosbag2_cpp/writer.hpp"
+#include "rosbag2_cpp/writers/sequential_writer.hpp"
+#include "rosbag2_storage/ros_helper.hpp"
+#include "rosbag2_storage/storage_filter.hpp"
+#include "rosbag2_storage/topic_metadata.hpp"
+
+namespace rosbag2_py
+{
+
+class Reader
+{
+public:
+  explicit Reader(const std::string & reader_class)
+  {
+    if (!reader_class.compare("SequentialReader")) {
+      auto sequential_reader = std::make_unique<rosbag2_cpp::readers::SequentialReader>();
+      reader_ = std::make_unique<rosbag2_cpp::Reader>(std::move(sequential_reader));
+    } else if (!reader_class.compare("SequentialCompressionReader")) {
+      auto sequential_reader = std::make_unique<rosbag2_compression::SequentialCompressionReader>();
+      reader_ = std::make_unique<rosbag2_cpp::Reader>(std::move(sequential_reader));
+    } else {
+      throw std::runtime_error{"Reader class type " + reader_class + " not supported."};
+    }
+  }
+
+  void open(
+    rosbag2_cpp::StorageOptions & storage_options,
+    rosbag2_cpp::ConverterOptions & converter_options)
+  {
+    reader_->open(storage_options, converter_options);
+  }
+
+  bool has_next()
+  {
+    return reader_->has_next();
+  }
+
+  /// Return a tuple containing the topic name, the serialized ROS message, and
+  /// the timestamp.
+  pybind11::tuple read_next()
+  {
+    const auto next = reader_->read_next();
+    rcutils_uint8_array_t rcutils_data = *next->serialized_data.get();
+    std::string serialized_data(rcutils_data.buffer,
+      rcutils_data.buffer + rcutils_data.buffer_length);
```

Comment:
why can't we return a uint8_t array? If we have to pack this into a STL container, I'd prefer `std::vector<uint8_t>`

Reply:
I asked for Jacob's help looking into this. When `read_next()` is called in Python, such as in the `test_sequential_reader.py`, the message needs to be deserialized. That was enabled in a new PR in `rclpy`, which takes a byte array as input: https://github.com/ros2/rclpy/pull/495/files#diff-80dda03110b5be606b823d6b0558b5c2R4745 . Currently `read_next()` returns a `pybind11::bytes`, which translates to a Python `bytes` type and can be passed directly to `rclpy.serialization.deserialize_message()`

I think it is possible to return a `std::vector<uint8_t>`, but that would translate to a python `list` type of ints. So it would require an extra step in the Python code on the user's end to convert to a `bytes` type in order to deserialize it.

Reply:
I think that's fine. I was more wondering about why using a `std::string` here for the serialized buffer.

Reply:
I think it's just that the conversion between string and bytes is the easiest and native to what Python [`bytes` / `bytearray`](https://docs.python.org/3/library/functions.html#func-bytearray) do.